In [ ]:
## Install the required libraries
%pip install requests
%pip install pandas
%pip install openpyxl

In [ ]:
## Loading libraries
import json
from urllib.request import urlopen
import csv
import pandas as pd
from datetime import datetime

## Task 1

### Part 1

In [ ]:
## Loading the json file and viewing the contents of the data

def load_json(url):
    with urlopen(url) as response:
        source = response.read()
    data = json.loads(source)
    print(json.dumps(data, indent=2))
    return data

url = "https://raw.githubusercontent.com/Papagoat/brain-assessment/main/restaurant_data.json"

data = load_json(url)


In [ ]:
## Understanding the structure of the data better
#len(data)
#len(data[0])
#len(data[0]["restaurants"])

In [ ]:
## Getting the Country Code dictionary from the Excel file
def country_codes(file):
    df = pd.read_excel(file)
    country_codes = dict(zip(df['Country Code'], df['Country'])) 
    return country_codes

country_codes = country_codes('Country-Code.xlsx')

In [ ]:
## Extracing the relevant fields

def get_restaurants(data):
    restaurants = []
    for i in data:
        for j in i["restaurants"]:
            country_id = j["restaurant"]["location"]["country_id"]
            if country_id in country_codes:
                country = country_codes[country_id]
            else:
                country = "NA"
            entry = [j["restaurant"]["id"], j["restaurant"]["name"], country, 
                    j["restaurant"]["location"]["city"], j["restaurant"]["user_rating"]["votes"], 
                    float(j["restaurant"]["user_rating"]["aggregate_rating"]), j["restaurant"]["cuisines"]]
            restaurants.append(entry)
    return restaurants
        
restaurants = get_restaurants(data)    

In [ ]:
## Coverting to csv file
def to_csv(headers, filename, data):
    with open(filename, "w", encoding="UTF-8", newline="") as file:
        writer = csv.writer(file)
        writer.writerow(headers)
        writer.writerows(data)

csvheader = ["RESTAURANT ID", "RESTAURANT NAME", "COUNTRY", "CITY",
                "USER RATING VOTES", "USER AGGREGATE RATING", "CUISINES"]

to_csv(csvheader, "restaurants.csv", restaurants)

### Part 2

In [ ]:
## Extracting the relevant fields for restaurants that have past events in the month of April 2019
def get_events(data):
    events = []

    april = pd.date_range(start='4/1/2019', end='4-30-2019').date
    format = '%Y-%m-%d'

    for i in data:
        for j in i["restaurants"]:

            if "zomato_events" in j["restaurant"]:
                for k in j["restaurant"]["zomato_events"]:
                    start = datetime.strptime(k["event"]["start_date"], format).date()
                    end = datetime.strptime(k["event"]["end_date"], format).date()
                    
                    if (start in april or end in april) or (start < april[0] and end > april[-1]):
                        url = ', \n'.join([m["photo"]["url"] for m in k["event"]["photos"]])
                        entry = [k["event"]["event_id"], j["restaurant"]["id"], j["restaurant"]["name"], 
                                url, k["event"]["title"], start, end]
                        events.append(entry)
    return events

events = get_events(data)

In [ ]:
## Coverting to csv file
csvheader = ["EVENT ID", "RESTAURANT ID", "RESTAURANT NAME", "PHOTO URL",
             "EVENT TITLE", "EVENT START DATE", "EVENT END DATE"]
def to_csv_na(headers, filename, data):
    with open(filename, "w", encoding="UTF-8", newline="") as file:
        writer = csv.writer(file)
        writer.writerow(csvheader)
        # populating empty cells with N.A.
        for event in data:
            event = [e if e != "" else 'N.A.' for e in event]
            writer.writerow(event)

to_csv_na(csvheader, "restaurants_events.csv", events)

### Part 3

In [57]:
## Getting the thresholds for different text ratings
def get_threshold(text, threshold):
    for i in data:
        for j in i["restaurants"]:
            rating_text = j["restaurant"]["user_rating"]["rating_text"]
            aggregate_rating = float(j["restaurant"]["user_rating"]["aggregate_rating"])
            if rating_text in text:
                if rating_text in threshold:
                    if aggregate_rating < threshold[rating_text]:
                        threshold[rating_text] = aggregate_rating
    print(threshold)

text = ["Excellent", "Very Good", "Good", "Average", "Poor"]
threshold = {"Excellent": 5.0, "Very Good": 5.0,"Good": 5.0, "Average": 5.0, "Poor": 5.0}

get_threshold(text, threshold)

{'Excellent': 4.5, 'Very Good': 4.0, 'Good': 3.5, 'Average': 2.5, 'Poor': 2.2}


However, assuming that the lowest rating is Poor, the threshold should be 0 instead. Thus, threshold of different rating text based on aggregate rating: 

Excellent: 4.5 ≤ x ≤ 5.0

Very Good: 4.0 ≤ x < 4.5

Good: 3.5 ≤ x < 4.0

Average: 2.5 ≤ x  < 3.5

Poor: x < 2.5